python/lsst/ap/association/packageAlerts.py

```diff
@@ -0,0 +1,247 @@
+        self.cutoutBBox = geom.Extent2I(self.config.cutoutSize,
+                                        self.config.cutoutSize)
+        os.makedirs(self.config.alertWriteLocation, exist_ok=True)
+
+    def run(self,
+            diaSourceCat,
+            diaObjectCat,
+            diaSrcHistory,
+            diffIm,
+            template,
+            ccdExposureIdBits):
+        """Package DiaSources/Object and exposure data into Avro alerts.
+
+        Writes Avro alerts to a location determined by the
+        ``alertWriteLocation`` configurable.
+
+        Parameters
+        ----------
+        diaSourceCat : `pandas.DataFrame`
+            New DiaSources to package. DataFrame should be indexed on
+            ``["diaObjectid", "filterName", "diaSourceId"]``
+        diaObjectCat : `pandas.DataFrame`
+            New and updated DiaObjects matched to the new DiaSources. DataFrame
+            is indexed on ``["diaObjectId"]``
+        diaSrcHistory : `pandas.DataFrame`
+            12 month history of DiaSources matched to the DiaObjects. Excludes
+            the newest DiaSource and is indexed on
+            ``["diaObjectid", "filterName", "diaSourceId"]``
+        diffIm : `lsst.afw.image.ExposureF`
+            Difference image ``diaSourceCat`` were detect in.
+        template : `lsst.afw.image.ExposureF` or `None`
+            Template image used to create the ``diffIm``.
+        ccdExposureIdBits : `int`
+            Number of bits used in the ccdVisitId.
+        """
+        alerts = []
+        self._patchDiaSources(diaSourceCat)
+        self._patchDiaSources(diaSrcHistory)
+        self._patchDiaObjects(diaObjectCat)
+        ccdVisitId = diffIm.getInfo().getVisitInfo().getExposureId()
+        for srcIndex, diaSource in diaSourceCat.iterrows():
+            # Get all diaSources for the associated diaObject.
+            diaObject = diaObjectCat.loc[srcIndex[0]]
+            if diaObject["nDiaSources"] > 1:
+                objSourceHistory = diaSrcHistory.loc[srcIndex[0]]
+            else:
+                objSourceHistory = None
+            sphPoint = geom.SpherePoint(diaSource["ra"],
+                                        diaSource["decl"],
+                                        geom.degrees)
+            diffImCutout = diffIm.getCutout(sphPoint, self.cutoutBBox)
+            templateCutout = None
+            alertId = diaSource["diaSourceId"]
+            alerts.append(
+                self.makeJsonAlert(alertId,
+                                   diaSource,
+                                   diaObject,
+                                   objSourceHistory,
+                                   diffImCutout,
+                                   templateCutout))
+        with open(os.path.join(self.config.alertWriteLocation,
+                               f"{ccdVisitId}.avro"),
+                  "ab+") as f:
+            self.alertSchema.store_alerts(f, alerts)
```

Comment:
Not sure how this would make it easier? The test I currently have for the run method would end up basically being the same and I'm already testing decomposing into the dict. Are you suggesting testing the list of alerts that are fully decomposed instead of just the single dict?

Reply:
So, here's what I'm thinking - but this is a judgement call, so feel free to say if you want to keep it as-is.

Suppose you split the function into two. `formatAlerts` makes the `alerts` list, and `writeAlerts` opens a file and calls `alertSchema.store_alerts`.

Your `testRun` currently does this:

```python
    def testRun(self):
        packConfig = PackageAlertsConfig()
        tempdir = tempfile.mkdtemp(prefix='alerts')
        packConfig.alertWriteLocation = tempdir
        packageAlerts = PackageAlertsTask(config=packConfig)

        packageAlerts.run(self.diaSources,
                          self.diaObjects,
                          self.diaSourceHistory,
                          self.exposure,
                          None,
                          None)

        ccdVisitId = self.exposure.getInfo().getVisitInfo().getExposureId()
        with open(os.path.join(tempdir, f"{ccdVisitId}.avro"), 'rb') as f:
            writer_schema, data = \
                packageAlerts.alertSchema.retrieve_alerts(f)

        # [ *actual* test code here]

        shutil.rmtree(tempdir)
```
All of this boilerplate is really just making it so the test can work with the filesystem. It's like `run` is a function whose return value can only be accessed by opening a file, so you have to do a bunch of work.

If `run` were split as I propose, this could be condensed way down for `testFormatAlerts`:
```python
    def testFormatAlerts(self):
        packConfig = PackageAlertsConfig()
        packageAlerts = PackageAlertsTask(config=packConfig)
        data = packageAlerts.formatAlerts(self.diaSources,
                                          self.diaObjects,
                                          self.diaSourceHistory,
                                          self.exposure,
                                          None,
                                          None)

        # [ *actual* test code here]
```

And separately you can `testWriteAlerts` if you want, but probably not even - it would pretty much only be testing that `lsst.alert.packet.write_alerts` and `lsst.alert.packet.retrieve_alerts` work. You usually don't need to test that other packages' code works, you want to test that _your_ code works and trust them to have their own tests.

Reply:
Hmm... I get you. I think testing the round tripped alert is still important as we want to be assured that the data sent through comes back within a defined tolerance. You could imagine a case where something in the Avro or the schema changes and precision is lost to an unacceptable level. With that in mind, I'll just leave it for now.